controllers/machine_controller_node.go

```diff
@@ -0,0 +1,166 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controllers
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/pkg/errors"
+	corev1 "k8s.io/api/core/v1"
+	clusterv1 "sigs.k8s.io/cluster-api/api/v1alpha3"
+	"sigs.k8s.io/cluster-api/controllers/noderefutil"
+	"sigs.k8s.io/cluster-api/util"
+	"sigs.k8s.io/cluster-api/util/conditions"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+)
+
+var (
+	ErrNodeNotFound = errors.New("cannot find node with matching ProviderID")
+)
+
+func (r *MachineReconciler) reconcileNode(ctx context.Context, cluster *clusterv1.Cluster, machine *clusterv1.Machine) (reconcile.Result, error) {
+	logger := r.Log.WithValues("machine", machine.Name, "namespace", machine.Namespace)
+
+	// Check that the Machine has a valid ProviderID.
+	if machine.Spec.ProviderID == nil || *machine.Spec.ProviderID == "" {
+		logger.Info("Cannot reconcile node, the machine doesn't have a valid ProviderID yet")
+		conditions.MarkFalse(machine, clusterv1.MachineNodeHealthyCondition, clusterv1.WaitingForNodeRefReason, clusterv1.ConditionSeverityInfo, "")
+		return reconcile.Result{}, nil
+	}
+
+	providerID, err := noderefutil.NewProviderID(*machine.Spec.ProviderID)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	remoteClient, err := r.Tracker.GetClient(ctx, util.ObjectKey(cluster))
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	// Even if Status.NodeRef exists, continue to do the following checks to make sure Node is healthy
+	node, err := r.getNode(remoteClient, providerID)
+	if err != nil {
+		if err == ErrNodeNotFound {
+			// While a NodeRef is set in the status, failing to get that node means the node is deleted.
```

Comment:
What if it's a transient error? The node might not be deleted

Reply:
We are setting the condition and reconciling, if it is transient it will be fixed right away. Otherwise how to detect the transient error?